graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/builtins/ArrayPrototypeBuiltins.java

```diff
@@ -3456,4 +3476,444 @@ protected Object createGroupResult(JSMapObject map, Map<Object, List<Object>> gr
+
+            for (tgtIdx = 0; tgtIdx < len; tgtIdx++) {
+                long srcIdx = len - tgtIdx - 1;
+                Object value = read(array, srcIdx);
+                JSRuntime.createDataPropertyOrThrow(resultArray, Strings.fromLong(tgtIdx), value);
+                TruffleSafepoint.poll(this);
+            }
+            reportLoopCount(tgtIdx);
+
+            return resultArray;
+        }
+    }
+
+    // TODO Schlaegl: Generalization: see spec (https://tc39.es/proposal-change-array-by-copy):
+    // "sort", "SortIndexedProperties", "CompareArrayElements")
+    public abstract static class JSArrayToSortedNode extends JSArrayOperation {
+
+        @Child private DeletePropertyNode deletePropertyNode; // DeletePropertyOrThrow
+        private final ConditionProfile isSparse = ConditionProfile.create();
+        private final BranchProfile hasCompareFnBranch = BranchProfile.create();
+        private final BranchProfile noCompareFnBranch = BranchProfile.create();
+        private final BranchProfile growProfile = BranchProfile.create();
+        @Child private InteropLibrary interopNode;
+        @Child private ImportValueNode importValueNode;
+
+        public JSArrayToSortedNode(JSContext context, JSBuiltin builtin, boolean isTypedArrayImplementation) {
+            super(context, builtin, isTypedArrayImplementation);
+        }
+
+        @Specialization(guards = {"!isTypedArrayImplementation", "isJSFastArray(thisObj)"}, assumptions = "getContext().getArrayPrototypeNoElementsAssumption()")
+        protected JSDynamicObject toSortedArray(final JSDynamicObject thisObj, final Object compare,
+                        @Cached("create(getContext())") JSArrayToDenseObjectArrayNode arrayToObjectArrayNode,
+                        @Cached("create(getContext(), true)") JSArrayDeleteRangeNode arrayDeleteRangeNode) {
+            checkCompareFunction(compare);
+            long len = getLength(thisObj);
+            ScriptArray scriptArray = arrayGetArrayType(thisObj);
+            Object[] array = arrayToObjectArrayNode.executeObjectArray(thisObj, scriptArray, len);
+
+            sortIntl(getComparator(thisObj, compare), array);
+            reportLoopCount(len); // best effort guess, let's not go for n*log(n)
+
+            JSDynamicObject resultArray = (JSDynamicObject) getArraySpeciesConstructorNode().arrayCreate(len);
+            for (int i = 0; i < array.length; i++) {
+                JSRuntime.createDataPropertyOrThrow(resultArray, Strings.fromLong(i), array[i]);
+            }
+
+            if (isSparse.profile(array.length < len)) {
+                arrayDeleteRangeNode.execute(resultArray, scriptArray, array.length, len);
+            }
+            return resultArray;
+        }
+
+        private void delete(Object obj, Object i) {
+            if (deletePropertyNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                JSContext context = getContext();
```

Comment:
(minor) unnecessary local variable, better have that inline. You use it only once; would be fine if you used it it 2+ times.